src/drivers/smart_battery/batmon/batmon.cpp

```diff
@@ -0,0 +1,352 @@
+ *
+ *   Copyright (c) 2021 PX4 Development Team. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file batmon.cpp
+ *
+ * BatMon module for Smart Battery utilizing SBS 1.1 specifications
+ * Setup/usage information: https://rotoye.com/batmon-tutorial/
+ *
+ * @author Eohan George <eohan@rotoye.com>
+ * @author Nick Belanger <nbelanger@mail.skymul.com>
+ */
+
+#include "batmon.h"
+#include <mathlib/mathlib.h>
+
+extern "C" __EXPORT int batmon_main(int argc, char *argv[]);
+
+Batmon::Batmon(I2CSPIBusOption bus_option, const int bus, SMBus *interface):
+	SMBUS_SBS_BaseClass(bus_option, bus, interface)
+{
+
+}
+
+Batmon::~Batmon()
+{
+	// Needn't change the parameter since there are other modules that may use it.
+	//int battsource = 0;
+	//param_set(param_find("BAT_SOURCE"), &battsource);
+}
+
+I2CSPIDriverBase *Batmon::instantiate(const BusCLIArguments &cli, const BusInstanceIterator &iterator,
+				      int runtime_instance)
+
+{
+	//TODO there may be a better way to pass the driver ID, since iterator object should have it
+	SMBus *interface = new SMBus(DRV_BAT_DEVTYPE_BATMON_SMBUS, iterator.bus(), cli.i2c_address);
+
+	int32_t batmon_en_param = 0;
+	param_get(param_find("BATMON_DRIVER_EN"), &batmon_en_param);
+
+	if (batmon_en_param == 0) {	// BATMON_DRIVER_EN is set to disabled. Do not start driver
+		return nullptr;        // TODO: add option for autodetect I2C address
+	}
```

Comment:
IMO it is redundant to check this here when it is already only started based on the param value.

Reply:
This is really a placeholder for other startup logic we'd like to add later based on the parameters. I understand that its redundant.